dash/include/dash/algorithm/Accumulate.h

```diff
@@ -86,45 +165,23 @@ ValueType accumulate(
 template <
   class GlobInputIt,
   class ValueType,
-  class BinaryOperation >
+  class BinaryOperation = dash::plus<ValueType>,
+  typename = typename std::enable_if<
+              std::is_same<decltype(std::declval<GlobInputIt>().team()),
+                           dash::Team&>::value>::type>
 ValueType accumulate(
-  GlobInputIt     in_first,
-  GlobInputIt     in_last,
-  ValueType       init,
-  BinaryOperation binary_op = dash::plus<ValueType>())
+        GlobInputIt   in_first,
+        GlobInputIt   in_last,
+  const ValueType   & init,
+  BinaryOperation     binary_op = dash::plus<ValueType>())
 {
-  struct local_result {
-    ValueType l_result;
-    bool      l_valid;
-  };
-
   auto & team      = in_first.team();
-  auto myid        = team.myid();
   auto index_range = dash::local_range(in_first, in_last);
   auto l_first     = index_range.begin;
   auto l_last      = index_range.end;
-  auto result      = init;
 
-  dash::Array<local_result> l_results(team.size(), team);
-  if (l_first != l_last) {
-    auto l_result  = std::accumulate(std::next(l_first), l_last, *l_first, binary_op);
-    l_results.local[0].l_result = l_result;
-    l_results.local[0].l_valid  = true;
-  } else {
-    l_results.local[0].l_valid  = false;
-  }
-
-  l_results.barrier();
-
-  if (myid == 0) {
-    for (size_t i = 0; i < team.size(); i++) {
-      local_result lr = l_results[i];
-      if (lr.l_valid) {
-        result = binary_op(result, lr.l_result);
-      }
-    }
-  }
-  return result;
+  // TODO: can we figure out whether or not units are empty?
```

Comment:
A bit verbose, maybe:
```c++
  auto empty = false;
  for (dash::team_unit_t unit{0}; unit < pattern.team().size(); unit++) {
     auto lsizes = pattern.local_extents(unit);
     empty = empty || std::any_of(lsizes.begin(), lsizes.end(), [](size_t size) { return size != 0; });
  }
```

Not sure if something like this is defined already. I couldn't find it.

Reply:
That would only account for the whole range of the of allocation, not the range defined by the iterators. The monkey in my brain yells *blocks blocks blocks* but I'm not sure they are of any help here.

Reply:
I have a similar problem in `dash::sort`. It is probably not the best approach but I have to know which units are empty.
See [here](https://github.com/dash-project/dash/blob/development/dash/include/dash/algorithm/Sort.h#L567)